internal/graph/check.go

```diff
@@ -575,16 +587,194 @@ func removeIndexFromSlice[T any](s []T, index int) []T {
+	ds := datastoremw.MustFromContext(ctx).SnapshotReader(crc.parentReq.Revision)
+	it, err := ds.QueryRelationships(ctx, datastore.RelationshipsFilter{
+		OptionalResourceType:     crc.parentReq.ResourceRelation.Namespace,
+		OptionalResourceIds:      crc.filteredResourceIDs,
+		OptionalResourceRelation: ttu.GetTupleset().GetRelation(),
+	})
+	if err != nil {
+		return checkResultError(NewCheckFailureErr(err), emptyMetadata)
+	}
+	defer it.Close()
+
+	subjectsToDispatch := tuple.NewONRByTypeSet()
+	relationshipsBySubjectONR := mapz.NewMultiMap[string, *core.RelationTuple]()
+	subjectsByResourceID := mapz.NewMultiMap[string, *core.ObjectAndRelation]()
+	for tpl := it.Next(); tpl != nil; tpl = it.Next() {
+		if it.Err() != nil {
+			return checkResultError(NewCheckFailureErr(it.Err()), emptyMetadata)
+		}
+
+		subjectsToDispatch.Add(tpl.Subject)
+		relationshipsBySubjectONR.Add(tuple.StringONR(tpl.Subject), tpl)
+		subjectsByResourceID.Add(tpl.ResourceAndRelation.ObjectId, tpl.Subject)
+	}
+	it.Close()
+
+	// Convert the subjects into batched requests.
+	// To simplify the logic, +1 is added to account for the situation where
+	// the number of elements is less than the chunk size, and spare us some annoying code.
+	expectedNumberOfChunks := uint16(subjectsToDispatch.ValueLen())/crc.maxDispatchCount + 1
+	toDispatch := make([]directDispatch, 0, expectedNumberOfChunks)
+	subjectsToDispatch.ForEachType(func(rr *core.RelationReference, resourceIds []string) {
+		chunkCount := 0.0
+		slicez.ForEachChunk(resourceIds, crc.maxDispatchCount, func(resourceIdChunk []string) {
+			chunkCount++
+			toDispatch = append(toDispatch, directDispatch{
+				resourceType: rr,
+				resourceIds:  resourceIdChunk,
+			})
+		})
+		dispatchChunkCountHistogram.Observe(chunkCount)
+	})
+
+	if subjectsToDispatch.IsEmpty() {
+		return noMembers()
+	}
+
+	// Run the dispatch for all the chunks. Unlike a standard TTU, we do *not* perform mapping here,
+	// as we need to access the results on a per subject basis. Instead, we keep each result and map
+	// by the relation type of the dispatched subject.
+	chunkResults, err := run(
+		ctx,
+		currentRequestContext{
+			parentReq:           crc.parentReq,
+			filteredResourceIDs: crc.filteredResourceIDs,
+			resultsSetting:      v1.DispatchCheckRequest_REQUIRE_ALL_RESULTS,
+			maxDispatchCount:    crc.maxDispatchCount,
+		},
+		toDispatch,
+		func(ctx context.Context, crc currentRequestContext, dd directDispatch) checkResultWithType {
+			childResult := cc.checkComputedUserset(ctx, crc, ttu.GetComputedUserset(), dd.resourceType, dd.resourceIds)
+			return checkResultWithType{
+				CheckResult:  childResult,
+				relationType: dd.resourceType,
+			}
+		},
+		cc.concurrencyLimit,
+	)
+	if err != nil {
+		return checkResultError(err, emptyMetadata)
+	}
+
+	// Create a membership set per-subject-type, representing the membership for each of the dispatched subjects.
+	resultsByDispatchedSubject := map[string]*MembershipSet{}
```

Comment:
I made a similar comment on one of the other graph methods, but you don't have to collect up all of the results before you combine them; you can combine as you get responses back.